pkg/apis/core/validation/validation_test.go

```diff
@@ -10540,13 +10546,15 @@ func TestValidateServiceCreate(t *testing.T) {
 			name: "valid type - loadbalancer",
 			tweakSvc: func(s *core.Service) {
 				s.Spec.Type = core.ServiceTypeLoadBalancer
+				s.Spec.AllocateLoadBalancerNodePorts = utilpointer.BoolPtr(true)
```

Comment:
We probably want a case where this is false

Reply:
Will add a negative case, thanks for the remainder.

Reply:
Added case where `AllocateLoadBalancerNodePorts` is set to `false` and another negative case that leaves `AllocateLoadBalancerNodePorts` unset.

Reply:
don't we have to enable the featuregate during the test?

```
	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceLBNodePortControl)()
```

Reply:
I think in this PR the feature is enabled by default already (see [here](https://github.com/kubernetes/kubernetes/blob/b1b8f049e7aa757d48559ffc2690271203b0cbf4/pkg/features/kube_features.go#L849)).

Reply:
hmm, I always thought that the test shouldn't depend on the global state, let me ask @liggitt , he always know all these things better than anybody.
Is just for learning purposes :smiley:

Reply:
This is validation though, so the feature gate shouldn't impact the logic?

Reply:
https://github.com/kubernetes/kubernetes/blob/204ff6caeb8229e6d12b54809108e7f32c19502f/pkg/apis/core/validation/validation.go#L4375-L4377

Reply:
My understanding is all features are loaded from `defaultKubernetesFeatureGates` to the global `DefaultMutableFeatureGate` in `init`https://github.com/kubernetes/kubernetes/blob/204ff6caeb8229e6d12b54809108e7f32c19502f/pkg/features/kube_features.go#L740-L742
And `ServiceLBNodePortControl` is now enabled in `defaultKubernetesFeatureGates`.
https://github.com/kubernetes/kubernetes/blob/b1b8f049e7aa757d48559ffc2690271203b0cbf4/pkg/features/kube_features.go#L849
Also `DefaultFeatureGate` is the same as `DefaultMutableFeatureGate`
https://github.com/kubernetes/kubernetes/blob/204ff6caeb8229e6d12b54809108e7f32c19502f/staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go#L30-L32

So later when the validation happens, it's using the global `DefaultFeatureGate`, so it should know that `ServiceLBNodePortControl` is being enabled.

Reply:
yeah, not a big deal, but my question was more related to know if we do want to test with and without feature gate ... but I think that I understand know Andrew's point, it will not be possible since we clear in strategy IIUIC
disregard my comment, sorry